### PR TITLE
[airbyte-cdk] - Integrate `HttpClient` into `HttpRequester`

### PR DESCRIPTION
**Pull request by pnilan**
## What
* Moves request preparation and sending functionality from `HttpRequester` to `HttpClient`
* Updates `HttpClient` for compatibility with low-code
    * Handles a list of backoff_strategies
    * Updates logic for determining `user_defined_backoff_time` to be compatible with list of backoff_strategie
* Updated `ErrorHandler` interface to enforce `max_retries` and `max_time` properties. 
* Updates `Requester` interface, removing `interpret_response` method. This logic is handled by the `ErrorHandler` class.

## How
Update is to align the low-code CDK interfaces with common Python CDK interfaces and to provide sensible error defaults for response actions and failure types for all low-code connectors.

Closes https://github.com/airbytehq/airbyte-internal-issues/issues/7864

## Recommended Reading Order
1. `http_client.py`
2. `model_to_component_factory`
3. `http_requester.py`
4. `requester.py`
5. `default_error_mapping`

### Notes
- Updated default error mapping to include InvalidSchema and InvalidURL exceptions

### CODE REVIEW

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py [hunk 1]
```diff
@@ -136,34 +147,92 @@ def _create_prepared_request(
...
+        log_formatter: Optional[Callable[[requests.Response], Any]] = None,
+    ) -> requests.Response:
 
         if request not in self._request_attempt_count:
             self._request_attempt_count[request] = 1
         else:
             self._request_attempt_count[request] += 1
 
-        self._logger.debug(
+        self._logger.info(
```
Comment (tolik0) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:235]: Why are you changing the log level? Logging all requests may be too noisy

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py [hunk 2]
```diff
@@ -208,9 +285,14 @@ def _send(self, request: requests.PreparedRequest, request_kwargs: Mapping[str,
 
         # TODO: Consider dynamic retry count depending on subsequent error codes
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:286]: Is this done as part of this PR?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py [hunk 3]
```diff
@@ -208,9 +285,14 @@ def _send(self, request: requests.PreparedRequest, request_kwargs: Mapping[str,
...
         elif error_resolution.response_action == ResponseAction.RETRY:
-            custom_backoff_time = self._backoff_strategy.backoff_time(
-                response_or_exception=response if response is not None else exc, attempt_count=self._request_attempt_count[request]
-            )
+            custom_backoff_time = None
+            for backoff_strategy in self._backoff_strategies:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:289]: This loop could make it so that `custom_backoff_time` is none. What happens in that case?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py [hunk 4]
```diff
@@ -136,34 +147,92 @@ def _create_prepared_request(
 
         return prepared_request
 
-    def _send_with_retry(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+    def _max_retries(self) -> int:
+        max_retries = self._DEFAULT_MAX_RETRY
+        if self._disable_retries:
+            max_retries = 0
+        elif hasattr(self._error_handler, "max_retries") and self._error_handler.max_retries is not None:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:154]: This condition seems very odd. Why do we have this here? Should we make it that everything relies on the interface `ErrorHandler`? 

The same comment applies to `max_time` and `retry_factor`

**Review by maxi297 — COMMENTED** [review]
Generally LGMT but the interfaces are a bit weird. We shouldn't have to `hasattr` or `isinstance`. 

Also, have we tested if this update impacts connectors already using HttpClient like Salesforce?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py [hunk 5]
```diff
@@ -136,34 +147,92 @@ def _create_prepared_request(
 
         return prepared_request
 
-    def _send_with_retry(self, request: requests.PreparedRequest, request_kwargs: Mapping[str, Any]) -> requests.Response:
+    def _max_retries(self) -> int:
```
Comment (tolik0) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:150]: Should we consider adding the `@cached_property` decorator to these methods?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 2 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:286]: No, I'll make a follow-up issue for this though.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 1 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:235]: Nice catch. This was a mistake.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 3 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:289]: If `custom_backoff_time` is None then the `DefaultBackoffException` is raised instead of `UserDefinedBackoffException`.

https://github.com/airbytehq/airbyte/blob/0b04c48f4660aed6e13cb11d152763ea880f783b/airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py#L300

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 4 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:154]: Yeah I recognize this is pretty weird.

Maybe I'm being too limiting, but the issue is that the existing low-code error handlers (e.g. `DefaultErrorHandler`) includes max retries and max_time, and the max_retries is actually specified by the connector dev in the manifest. The existing low-code backoff handlers (e.g. `ExponentialBackoffStrategy`) may include retry factor.

In the Python CDK, the `DefaultBackoffStrategy` includes max_retries and max_time.

In other words, these properties aren't enforced by the common interface. In order to minimize impact to existing low-code components (specifically because the error handler schema enables connector dev to specify a max_retry count), and because multiple backoff strategies can be specified for low-code. I opted to try to cater to both low-code and python implementations.

Let me know if you think it would be preferred to align more closely.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 3 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:289]: You are right. Should this be named `user_defined_backoff_time` then to align with the other concepts?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 4 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:154]: This somewhat makes sense but I'm still sceptic because:
* It feels like we are creating this brand new shiny class and it's already legacy because we purposefully decide not to respect the interfaces of our classes
* If I'm doing some development on the declarative package and changing the `max_retries` interface, now this will be broken without having some clear dependencies (the typing does not reflect that)
* It create weird cases where if I define an error_handler AND a backoff strategy that both have a max_retries, then the case is not clear for the user of this class

It feels like an easy problem to solve though: `HttpRequester` should instantiate proper implementations of `airbyte_cdk.sources.streams.http.error_handlers.BackoffStrategy`. What is preventing us from doing that?

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 3 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:289]: Yeah, I agree that makes sense.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 5 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:150]: Good idea! Added.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 4 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py]: @maxi297 Alright I agree with you.

I've update the ErrorHandler interface to enforce max_retries, max_time and aligned the `HttpStatusErrorHandler` with the updated interface.

Now all ErrorHandler implementations will have max_retries and max_time.

As for `factor`, that is still only available included for the `ExponentialBackoffStrategy` (or any custom backoff strategy that wants to implement a custom backoff strategy, so I kept the `hasattr` check for that.

Updated accordingly.

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py [hunk 6]
```diff
@@ -4,34 +4,23 @@
...
 from airbyte_cdk.sources.declarative.requesters.requester import HttpMethod, Requester
-from airbyte_cdk.sources.http_config import MAX_CONNECTION_POOL_SIZE
 from airbyte_cdk.sources.message import MessageRepository, NoopMessageRepository
-from airbyte_cdk.sources.streams.http.error_handlers import ErrorResolution, ResponseAction
-from airbyte_cdk.sources.streams.http.error_handlers.error_handler import ErrorHandler
-from airbyte_cdk.sources.streams.http.exceptions import DefaultBackoffException, RequestBodyException, UserDefinedBackoffException
-from airbyte_cdk.sources.streams.http.http import BODY_REQUEST_METHODS
-from airbyte_cdk.sources.streams.http.rate_limiting import http_client_default_backoff_handler, user_defined_backoff_handler
+from airbyte_cdk.sources.streams.http import HttpClient
+from airbyte_cdk.sources.streams.http.error_handlers import ErrorHandler
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py:21]: I'm not sure I get the typing here. It feels like what is being passed to `HttpRequested` from `model_to_component_factory` is `airbyte_cdk.sources.declarative.requesters.error_handlers.error_handler.ErrorHandler`. However, this is another error handler entirely. Is MyPy complaining about the types not matching? It feels like it's only working because we've made the interfaces matches but if there is a change in one, the other might not be updated though.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py [hunk 7]
```diff
@@ -136,27 +148,75 @@ def _create_prepared_request(
...
+
+    @cached_property
+    def _max_time(self) -> int:
+        return self._error_handler.max_time if self._error_handler.max_time is not None else self._DEFAULT_MAX_TIME
+
+    @cached_property
+    def _factor(self) -> float:
+        factor: Optional[float] = None
+        for backoff_strategy in self._backoff_strategies:
+            if hasattr(backoff_strategy, "retry_factor") and backoff_strategy.retry_factor is not None:
```
Comment (maxi297) [inline airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:168]: It seems like we are still relying on the declarative interfaces here. Should we have this `factor` as part of the `airbyte_cdk.sources.streams.http` package?

**Review by maxi297 — COMMENTED** [review]
I think there are still issues with the interface. Can you make sure this is not me missing something here and reply to my comments?

File: airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py (hunk 6 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py:21]: Following up for visibility: Maxime and I met and we realized the disconnect was because I had made changes in this PR: https://github.com/airbytehq/airbyte/pull/38743 which make the updates to the declarative interface and implementations.

The declarative backoffs, error_handlers now inherit from the Python CDK interfaces.

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 7 above)
Comment (pnilan) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:168]: Follow-up comment: The hasattr exists here because retry_factor is not enforced by the interface and is only used by exponential backoffs.

File: airbyte-cdk/python/unit_tests/connector_builder/test_connector_builder_handler.py [hunk 8]
```diff
@@ -657,10 +657,22 @@ def test_create_source():
 
     source = create_source(config, limits)
 
+    max_retries = 0
```
Comment (maxi297) [inline airbyte-cdk/python/unit_tests/connector_builder/test_connector_builder_handler.py:660]: Is it safe enough to rely on `assert http_client._disable_retries`? Else, it feels like we are copy/pasting the logic of 
`HttpClient. _max_retries` which means potentially duplication of work every time the logic change as this will need to be updated

File: airbyte-cdk/python/unit_tests/sources/streams/http/test_http_client.py [hunk 9]
```diff
@@ -391,3 +401,57 @@ def update_response(*args, **kwargs):
 
     assert http_client._session.send.call_count == call_count
     assert returned_response == valid_response
+
+
+def test_disable_retries():
```
Comment (maxi297) [inline airbyte-cdk/python/unit_tests/sources/streams/http/test_http_client.py:406]: These tests assert on private parameters. If we change the internal logic without changing the behavior, those will break. Should we instead do a test with requests_mock that ensure that we perform only one query in case of failure? The same question applies to the tests below

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 5 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:152]: That assumes that `self._error_handler.max_retries` is a constant in our current implementations. I think it is but I just wanted to point that out

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 7 above)
Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:168]: What it means to be is that we got our abstractions wrong and we are at risk of a breaking change once we have to update this concept. 

The easy solution I see is to have `BackoffHandler` implement `def handle(?) -> Callable[[request, request_kwargs, log_formatter], Response]` or something like that and stop exposing  `max_tries`, `max_time` and `factor`. This though is a bigger change today. I'm fine with pushing this breaking change for later assuming the risk of occurrence is low and the blast radius is somewhat contained

**Review by maxi297 — COMMENTED** [review]
I would like the approval from someone that has more context (I'm not aware of all the changes here) but from my perspective, there is just the testing to update

File: airbyte-cdk/python/unit_tests/connector_builder/test_connector_builder_handler.py (hunk 8 above)
Comment (pnilan) [reply airbyte-cdk/python/unit_tests/connector_builder/test_connector_builder_handler.py:660]: Good catch. I added in that logic before consolidating it in the http_client._max_retries method.

Comment (pnilan) [reply airbyte-cdk/python/unit_tests/connector_builder/test_connector_builder_handler.py:660]: Honestly I don't think this assertion is even worth keeping around since it's already tested via http_client.

Comment (maxi297) [reply airbyte-cdk/python/unit_tests/connector_builder/test_connector_builder_handler.py:660]: I don't think this is enough. We need to make sure the Connector Builder disable retries too

Comment (pnilan) [reply airbyte-cdk/python/unit_tests/connector_builder/test_connector_builder_handler.py]: Yeah, fair point. Added a separate test to specifically test for this.

File: airbyte-cdk/python/unit_tests/sources/streams/http/test_http_client.py (hunk 9 above)
Comment (pnilan) [reply airbyte-cdk/python/unit_tests/sources/streams/http/test_http_client.py:406]: Mocked

**Review by tolik0 — APPROVED** [review]

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 5 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:152]: this assumes all implementations of `max_retries` are constant. Since ErrorHandler is an interface that can be customized by any connector, I don't think we can make that assumption

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 7 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:168]: I agree with Maxime that the abstraction looks off. Why can't we rely on the backoff strategie's `backoff_time` to determine how long to wait before retrying? 

The declarative framework already supports exponential backoff without needing to leak the factor abstraction to the requester so we should be able to do the same here

**Review by girarda — CHANGES_REQUESTED** [review]
one small comment on the cached property and a larger question about why the `factor` concept leaks from the exponential error handling to the http client. That's not necessary in the existing requester so it should be necessary in the client either

File: airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py (hunk 7 above)
Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:168]: womp. looking at the http requester again, I see that the factor is indeed leaking because of the default value https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/requesters/http_requester.py#L504C9-L504C24

Comment (maxi297) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:168]: I think it's more than default value here: it is the fact that we can have multiple backoff strategies that relies on different parameters that we expose. Hence, the implementation of the backoff strategy are not well encapsulated. We could encapsulate that in the HttpClient if we wanted

Comment (girarda) [reply airbyte-cdk/python/airbyte_cdk/sources/streams/http/http_client.py:168]: @tolik0 @pnilan is there a reason we couldn't just use the [user_backoff_handler](https://github.com/airbytehq/airbyte/pull/38906/files#diff-6082346b7f2b198151e1a7b4629dd8428ab9330d225667b3dc5259e39a48434eR208)?

I wouldn't expect the default to be required since a backoff strategy should always be defined

**Comment by pnilan** [comment]
> one small comment on the cached property and a larger question about why the `factor` concept leaks from the exponential error handling to the http client. That's not necessary in the existing requester so it should be necessary in the client either

Removed reference to the `factor` property within http_client. Updated the `ExponentialBackoffStrategy` to only use this value internally to calc the backoff time and return it. The backoff handler has a default factor value of 1.

**Review by girarda — APPROVED** [review]
this is great! good job @pnilan !
